cmd/xl-storage.go

```diff
@@ -893,18 +893,26 @@ func (s *xlStorage) DeleteVersions(ctx context.Context, volume string, versions
 	return errs
 }
 
-// DeleteVersion - deletes FileInfo metadata for path at `xl.meta`
+// DeleteVersion - deletes FileInfo metadata for path at `xl.meta`. Create a fresh
+// `xl.meta` if it does not exist and we creating a new delete-marker.
 func (s *xlStorage) DeleteVersion(ctx context.Context, volume, path string, fi FileInfo) error {
 	if HasSuffix(path, SlashSeparator) {
 		return s.Delete(ctx, volume, path, false)
 	}
 
 	buf, err := s.ReadAll(ctx, volume, pathJoin(path, xlStorageFormatFile))
 	if err != nil {
-		if err == errFileNotFound && fi.VersionID != "" {
-			err = errFileVersionNotFound
+		if err != errFileNotFound {
+			return err
 		}
-		return err
+		if fi.Deleted {
+			// Create a new xl.meta with a delete marker in it
+			return s.WriteMetadata(ctx, volume, path, fi)
+		}
+		if fi.VersionID != "" {
+			return errFileVersionNotFound
+		}
+		return errFileNotFound
```

Comment:
Would we not create a marker here if the bucket doesn't have versioning enabled? @vadmeste

Reply:
> Would we not create a marker here if the bucket doesn't have versioning enabled? @vadmeste

yeah I thought about it, if the bucket is not versioning, then we will never have fi.Deleted (a delete marker) to be true.

Reply:
Can we make a bit for strict than just this ? @vadmeste because explicit check is better than relying on fi.Deleted

Reply:
main reason is if this call gets used in healing, healing might create random delete markers for no reason. We should also test this with replication feature on how delete marker would get replicated.

@poornas @vadmeste

Reply:
> Can we make a bit for strict than just this ? @vadmeste because explicit check is better than relying on fi.Deleted

By the way, this PR still won't create delete marker when deleting a non existant object (just tested again). What it does is, it creates a delete marker using minio-go Internal Delete API and I needed that for mc mirror replication.

Reply:
yes but these are extremely intricate details and we ought to sufficiently document them @vadmeste in the code two ways to do it 

- document heavily this region on why it is done this way
- or add an explicit value `fi.Versioned`  indicating if the bucket is versioned to perform this operation

Reply:
I think this is an overkill @harshavardhana, 

```
       // Deleted is set when this FileInfo represents
        // a deleted marker for a versioned bucket.
        Deleted bool
```

IMO, this is good enough, Deleted is well defined. If something goes wrong, it is the caller fault not xl-storage.go

Reply:
Ah okay fair enough @vadmeste